core/commands/p2p.go

```diff
@@ -53,124 +61,186 @@ are refined`,
+		}
+
+		if err := forwardLocal(n.Context(), n.P2P, n.Peerstore, proto, listen, target); err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+		res.SetOutput(nil)
 	},
 }
 
-var p2pListenerLsCmd = &cmds.Command{
+var p2pListenCmd = &cmds.Command{
 	Helptext: cmdkit.HelpText{
-		Tagline: "List active p2p listeners.",
+		Tagline: "Create libp2p service",
+		ShortDescription: `
+Create libp2p service and forward connections made to <target-address>.
+
+<protocol> specifies the libp2p handler name. It must be prefixed with '` + P2PProtoPrefix + `'.
+
+Example:
+  ipfs p2p listen ` + P2PProtoPrefix + `myproto /ip4/127.0.0.1/tcp/1234
+    - Forward connections to 'myproto' libp2p service to 127.0.0.1:1234
+
+`,
+	},
+	Arguments: []cmdkit.Argument{
+		cmdkit.StringArg("protocol", true, false, "Protocol name."),
+		cmdkit.StringArg("target-address", true, false, "Target endpoint."),
 	},
 	Options: []cmdkit.Option{
-		cmdkit.BoolOption("headers", "v", "Print table headers (HandlerID, Protocol, Local, Remote)."),
+		cmdkit.BoolOption("allow-custom-protocol", "Don't require /x/ prefix"),
 	},
 	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := p2pGetNode(req)
+		if err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+
+		protoOpt := req.Arguments()[0]
+		targetOpt := req.Arguments()[1]
+
+		proto := protocol.ID(protoOpt)
 
-		n, err := getNode(req)
+		target, err := ma.NewMultiaddr(targetOpt)
 		if err != nil {
 			res.SetError(err, cmdkit.ErrNormal)
 			return
 		}
 
-		output := &P2PLsOutput{}
+		allowCustom, _, err := req.Option("allow-custom-protocol").Bool()
+		if err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
 
-		for _, listener := range n.P2P.Listeners.Listeners {
-			output.Listeners = append(output.Listeners, P2PListenerInfoOutput{
-				Protocol: listener.Protocol,
-				Address:  listener.Address.String(),
-			})
+		if !allowCustom && !strings.HasPrefix(string(proto), P2PProtoPrefix) {
+			res.SetError(errors.New("protocol name must be within '"+P2PProtoPrefix+"' namespace"), cmdkit.ErrNormal)
+			return
 		}
 
-		res.SetOutput(output)
+		if err := forwardRemote(n.Context(), n.P2P, proto, target); err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+
+		res.SetOutput(nil)
 	},
-	Type: P2PLsOutput{},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: func(res cmds.Response) (io.Reader, error) {
-			v, err := unwrapOutput(res.Output())
-			if err != nil {
-				return nil, err
-			}
+}
 
-			headers, _, _ := res.Request().Option("headers").Bool()
-			list := v.(*P2PLsOutput)
-			buf := new(bytes.Buffer)
-			w := tabwriter.NewWriter(buf, 1, 2, 1, ' ', 0)
-			for _, listener := range list.Listeners {
-				if headers {
-					fmt.Fprintln(w, "Address\tProtocol")
-				}
+// forwardRemote forwards libp2p service connections to a manet address
+func forwardRemote(ctx context.Context, p *p2p.P2P, proto protocol.ID, target ma.Multiaddr) error {
+	// TODO: return some info
```

Comment:
well, the 'handle' here is really just what the user specified. In the forwardLocal case this makes a bit more sense in case where user specifies tcp port 0, where we'd return what the OS picked.

Reply:
Hm. Right, they can just shut it down via the protocol name.